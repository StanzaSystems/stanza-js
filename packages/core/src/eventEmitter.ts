class StanzaChangeEvent<StanzaEvent> extends Event {
  public readonly detail: StanzaEvent

  constructor (state: StanzaEvent) {
    super('stanzaStateChanged')
    this.detail = state
  }
}

type StanzaListener<StanzaEvent, ListenerReturnValue = unknown> = (event: StanzaEvent) => ListenerReturnValue

export interface StanzaChangeEmitter<StanzaEvent> {
  addChangeListener: (callback: StanzaListener<StanzaEvent>) => () => void
  removeChangeListener: (callback: StanzaListener<StanzaEvent>) => void
  dispatchChange: (state: StanzaEvent) => void
}

export class StanzaChangeTarget<StanzaEvent> implements StanzaChangeEmitter<StanzaEvent> {
  private readonly eventTarget = new EventTarget()
  private readonly unsubscribeMap = new Map<StanzaListener<StanzaEvent>, () => void>()

  addChangeListener (callback: StanzaListener<StanzaEvent>): () => void {
    const eventListener = (event: unknown): void => {
      if (!(event instanceof StanzaChangeEvent)) {
        return
      }
      callback(event.detail)
    }

    const unsubscribe = (): void => {
      this.unsubscribeMap.delete(callback)
      this.eventTarget.removeEventListener('stanzaStateChanged', eventListener)
    }

    this.unsubscribeMap.set(callback, unsubscribe)
    this.eventTarget.addEventListener('stanzaStateChanged', eventListener)

    return unsubscribe
  }

  removeChangeListener (callback: StanzaListener<StanzaEvent>): void {
    const unsubscribe = this.unsubscribeMap.get(callback)

    if (typeof unsubscribe === 'function') {
      unsubscribe()
    }
  }

  dispatchChange (state: StanzaEvent): void {
    this.eventTarget.dispatchEvent(new StanzaChangeEvent(state))
  }
}
