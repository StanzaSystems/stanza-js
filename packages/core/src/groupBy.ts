type MapValueFn<V, R> = (value: V) => R

export function groupBy<V extends Record<K, string>, K extends string, R = V> (key: K, mapValue: MapValueFn<V, R>) {
  return (grouped: Record<V[K], R>, value: V) => {
    const valueElement = value[key]
    grouped[valueElement] = mapValue(value)
    return grouped
  }
}
