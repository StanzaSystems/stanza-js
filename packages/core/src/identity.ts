export const identity = <V>(value: V): V => value;
