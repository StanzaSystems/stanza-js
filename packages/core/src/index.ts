import type { LocalStateProvider } from './models/LocalStateProvider'
import type { StanzaConfig } from './models/StanzaConfig'
import { createStanzaState, type StanzaState } from './models/StanzaState'
import { type Feature, FeatureStatus } from './models/Feature'
import InMemoryLocalStateProvider from './utils/InMemoryLocalStateProvider'

let stateProvider: LocalStateProvider
let stanzaConfig: StanzaConfig

const init = (config: StanzaConfig, provider?: LocalStateProvider): void => {
  if (stanzaConfig !== undefined || stateProvider !== undefined) {
    throw new Error('Stanza already initialized')
  }
  stateProvider = (provider != null) ? provider : InMemoryLocalStateProvider
  stanzaConfig = config
}

async function getRefreshStateForPageFeatures (page: string): Promise<StanzaState> {
  interface JSONResponse {
    data?: {
      features: Feature[]
    }
  }
  const params = new URLSearchParams()
  stanzaConfig.PageFeatures.get(page)?.forEach(s => { params.append('feature', s) })
  const response = await fetch(`${stanzaConfig.Url}/featureStatus?${params.toString()}`, {
    headers: {
      'x-stanza-customer-id': stanzaConfig.StanzaCustomerId
    }
  })
  const { data }: JSONResponse = await response.json()
  return createStanzaState(data?.features ?? [], page)
}

function savePageState (pageState: StanzaState): void {
  const pageConfig = stanzaConfig.PageFeatures.get(pageState.Page)
  if (pageConfig === null || pageConfig === undefined) {
    throw new Error(`configruation for page ${pageState.Page} not found`)
  }
  stateProvider.SetState(pageState, pageState.Page)
}

function getPageState (page: string): StanzaState {
  const state = stateProvider.GetState(page)
  if (state === undefined) {
    throw new Error(`State for page ${page} is not found. Is it configured?`)
  }
  return state
}

export { init, getRefreshStateForPageFeatures, savePageState, getPageState, FeatureStatus, stanzaConfig }
export type { StanzaState, StanzaConfig, LocalStateProvider }
