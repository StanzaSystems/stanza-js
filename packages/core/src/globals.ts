import { StanzaChangeTarget } from './eventEmitter';
import { groupBy } from './groupBy';
import { type FeatureState } from './models/featureState';
import { type LocalStateProvider } from './models/localStateProvider';
import { type StanzaCoreConfig } from './models/stanzaCoreConfig';

interface StanzaInternalConfig {
  environment: string;
  stanzaApiKey: string;
  url: string;
  refreshSeconds?: number;
  enablementNumberGenerator: () => Promise<number>;
  contextConfigs: Record<string, { features: string[] }>;
}

let stanzaConfig: StanzaInternalConfig;
let localStateProvider: LocalStateProvider;
let enablementNumber = 100;

export const featureChanges = new StanzaChangeTarget<FeatureState>();
export const enablementNumberChanges = new StanzaChangeTarget<number>();

export function init(
  config: StanzaCoreConfig,
  provider: LocalStateProvider,
): void {
  if (stanzaConfig !== undefined || localStateProvider !== undefined) {
    throw new Error('Stanza is already initialized');
  }
  stanzaConfig = {
    ...config,
    enablementNumberGenerator:
      config.enablementNumberGenerator ?? getEnablementNumberSimple,
    contextConfigs: config.contextConfigs.reduce(
      groupBy('name', ({ features }) => ({ features })),
      {},
    ),
  };
  localStateProvider = provider;

  localStateProvider.init(config.contextConfigs);

  getEnablementNumber().catch((e) => {
    console.warn('Failed to get enablement number', e);
  });
}

export function getConfig(): StanzaInternalConfig {
  if (stanzaConfig === undefined) {
    throw new Error('Stanza is not initialized');
  }
  return stanzaConfig;
}

export function getStateProvider(): LocalStateProvider {
  if (localStateProvider === undefined) {
    throw new Error('Stanza is not initialized');
  }
  return localStateProvider;
}

export async function getEnablementNumber(): Promise<number> {
  return stanzaConfig.enablementNumberGenerator().then((nr) => {
    enablementNumber = nr;
    enablementNumberChanges.dispatchChange(nr);
    return nr;
  });
}

export function getEnablementNumberStale(): number {
  return enablementNumber;
}

async function getEnablementNumberSimple(): Promise<number> {
  return Promise.resolve(Math.floor(Math.random() * 99));
}
