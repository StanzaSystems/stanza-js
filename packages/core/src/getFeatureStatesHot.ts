import { getStateProvider } from './globals'
import { type FeatureState } from './models/featureState'
import { fetchFeatureStates } from './utils/fetchFeatureStates'

export async function getFeatureStatesHot (features: string[]): Promise<FeatureState[]> {
  const featureStates = await fetchFeatureStates(features)
  const stateProvider = getStateProvider()
  featureStates.forEach(featureState => {
    stateProvider.setFeatureState(featureState)
  })

  return featureStates
}
