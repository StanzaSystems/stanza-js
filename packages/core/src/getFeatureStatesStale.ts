import { getStateProvider } from './globals';
import { createFeatureState } from './models/createFeatureState';
import { type FeatureState } from './models/featureState';
import {
  type AsyncLocalStateProvider,
  type LocalStateProvider,
} from './models/localStateProvider';

export function getFeatureStatesStale(features: string[]): FeatureState[] {
  const stateProvider = getStateProvider() as LocalStateProvider;
  const featureStates = features.map(
    (name) => stateProvider.getFeatureState(name) ?? createFeatureState(name)
  );
  featureStates.forEach((featureState) => {
    stateProvider.setFeatureState(featureState);
  });
  return featureStates;
}

export async function getFeatureStatesStaleAsync(
  features: string[]
): Promise<FeatureState[]> {
  const stateProvider = getStateProvider() as AsyncLocalStateProvider;
  const featureStates = await Promise.all(
    features.map(
      async (name) =>
        (await stateProvider.getFeatureState(name)) ?? createFeatureState(name)
    )
  );
  await Promise.all(
    featureStates.map(async (featureState) =>
      stateProvider.setFeatureState(featureState)
    )
  );

  return featureStates;
}
