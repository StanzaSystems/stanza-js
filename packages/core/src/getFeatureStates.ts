import { getFeatureStatesHot } from './getFeatureStatesHot';
import { getFeatureStatesStale } from './getFeatureStatesStale';
import { type FeatureState } from './models/featureState';
import { isFeatureStateFresh } from './models/isFeatureStateFresh';

export async function getFeatureStates(
  features: string[],
): Promise<FeatureState[]> {
  const featureStates = getFeatureStatesStale(features);
  if (
    featureStates.every(isFeatureStateFresh) &&
    features.every(
      (featureName) =>
        featureStates.findIndex(
          (state) => state.featureName === featureName,
        ) !== -1,
    )
  ) {
    return featureStates;
  }
  return getFeatureStatesHot(features);
}
