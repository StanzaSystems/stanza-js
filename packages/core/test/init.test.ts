import { assert, describe, expect, it } from 'vitest'
import * as Stanza from '../src/index'
import type { StanzaConfig } from '../src/index'

describe('init stanza', () => {
  it('configures a stanza instance', () => {
    const config: StanzaConfig = {
      Url: 'http://localhost:3004',
      Environment: 'local',
      StanzaCustomerId: '12345667',
      PageFeatures: new Map<string, string[]>(Object.entries({
        main: ['featured', 'search', 'checkout'],
        details: ['productSummary', 'pricing', 'shipping', 'checkout']
      }))
    }

    expect(() => { Stanza.init(config) }).to.not.throw()
    assert.deepEqual(Stanza.stanzaConfig, config, 'config is unchanged on init')
  })

  it('configures only one stanza', () => {
    const config: StanzaConfig = {
      Url: 'http://localhost:3004',
      Environment: 'local',
      StanzaCustomerId: '12345667',
      PageFeatures: new Map<string, string[]>(Object.entries({
        main: ['featured', 'search', 'checkout'],
        details: ['productSummary', 'pricing', 'shipping', 'checkout']
      }))
    }
    expect(() => { Stanza.init(config) }).to.throw()
  })
})
