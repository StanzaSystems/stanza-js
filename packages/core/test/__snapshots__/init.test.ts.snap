// Vitest Snapshot v1

exports[`init stanza > snapshot 1`] = `
{
  "foo": "bar",
}
`;

exports[`suite name > snapshot 1`] = `
{
  "foo": "bar",
}
`;
